source/material_model/steinberger.cc

```diff
@@ -578,71 +454,86 @@ namespace aspect
 
-      if (latent_heat)
-        {
-          /* We separate the calculation of specific heat and thermal expansivity,
-           * because they may depend on cell-wise averaged values that are only
-           * available here.
-           */
-          double average_temperature(0.0);
-          double average_density(0.0);
-          for (unsigned int i = 0; i < in.n_evaluation_points(); ++i)
-            {
-              average_temperature += in.temperature[i];
-              average_density += out.densities[i];
-            }
-          average_temperature /= in.n_evaluation_points();
-          average_density /= in.n_evaluation_points();
+          // The second derivatives of the thermodynamic potentials (compressibility, thermal expansivity, specific heat)
+          // are dependent not only on the phases present in the assemblage at the given temperature and pressure,
+          // but also on any reactions between phases in the assemblage. PerpleX and HeFESTo output only "static" properties
+          // (properties not including any reaction effects), and so do not capture the latent heat of reaction.
 
-          std::array<std::pair<double, unsigned int>,2> dH;
-          if (in.current_cell.state() == IteratorState::valid)
-            dH = enthalpy_derivative(in);
+          // In this material model, we always use a compressibility which includes the effects of reaction,
+          // but we allow the user the option to switch on or off thermal (latent heat) effects.
+          // If the latent_heat bool is set to true, thermal expansivity and specific heat are calculated from
+          // the change in enthalpy with pressure and temperature.
 
-          for (unsigned int i = 0; i < in.n_evaluation_points(); ++i)
+          // There are alternative ways to capture the latent heat effect (by preprocessing the P-T table, for example),
+          // which may be a more appropriate approach in some cases, but the latent heat should always be considered if
+          // thermodynamic self-consistency is intended.
+          if (latent_heat)
             {
-              // Use the adiabatic pressure instead of the real one,
-              // to stabilize against pressure oscillations in phase transitions
-              const double pressure = this->get_adiabatic_conditions().pressure(in.position[i]);
-
-              // If all of the derivatives were computed successfully
-              if ((in.current_cell.state() == IteratorState::valid)
-                  && (dH[0].second > 0)
-                  && (dH[1].second > 0))
+              if (this->get_adiabatic_conditions().is_initialized() && (in.current_cell.state() == IteratorState::valid)
+                  && (dH[0].second > 0) && (dH[1].second > 0))
                 {
-                  // alpha = (1 - rho * dH/dp) / T
                   out.thermal_expansion_coefficients[i] = (1 - average_density * dH[1].first) / average_temperature;
-                  // cp = dH/dT
```

Comment:
I would like to keep these comments, mostly to explain what `dH[1].first` and `dH[0].first` means. Feel free to reword, but keep that explanation somewhere

Reply:
Good point. I've moved the comments to the start of the if(latent_heat) clause.